eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/OfferCodecs.scala

```diff
@@ -0,0 +1,227 @@
+    .typecase(UInt64(38), payerKey)
+    .typecase(UInt64(39), payerNote)
+    .typecase(UInt64(50), payerInfo)
+    .typecase(UInt64(56), replaceInvoice)
+    .typecase(UInt64(240), signature)).complete
+
+  val invoiceRequestCodec: Codec[InvoiceRequest] = invoiceRequestTlvCodec.narrow({ tlvs =>
+    if (tlvs.get[OfferId].isEmpty) {
+      Attempt.failure(MissingRequiredTlv(UInt64(4)))
+    } else if (tlvs.get[PayerKey].isEmpty) {
+      Attempt.failure(MissingRequiredTlv(UInt64(38)))
+    } else if (tlvs.get[Signature].isEmpty) {
+      Attempt.failure(MissingRequiredTlv(UInt64(240)))
+    } else {
+      Attempt.successful(InvoiceRequest(tlvs))
+    }
+  }, {
+    case InvoiceRequest(tlvs) => tlvs
+  })
+
+  private val paymentInfo: Codec[PaymentInfo] = (("fee_base_msat" | millisatoshi32) ::
+      ("fee_proportional_millionths" | tu32) ::
+      ("cltv_expiry_delta" | cltvExpiryDelta)).as[PaymentInfo]
+
+  private val paymentPathsInfo: Codec[PaymentPathsInfo] = variableSizeBytesLong(varintoverflow, list(paymentInfo)).xmap[Seq[PaymentInfo]](_.toSeq, _.toList).as[PaymentPathsInfo]
+
+  private val paymentConstraints: Codec[PaymentConstraints] = (("max_cltv_expiry" | cltvExpiry) ::
+      ("htlc_minimum_msat" | millisatoshi) ::
+      ("allowed_features" | bytes.xmap[Features[Feature]](Features(_), _.toByteVector))).as[PaymentConstraints]
+
+  private val paymentPathsConstraints: Codec[PaymentPathsConstraints] = variableSizeBytesLong(varintoverflow, list(paymentConstraints)).xmap[Seq[PaymentConstraints]](_.toSeq, _.toList).as[PaymentPathsConstraints]
+
+  private val createdAt: Codec[CreatedAt] = variableSizeBytesLong(varintoverflow, tu64overflow).as[TimestampSecond].as[CreatedAt]
+
+  private val paymentHash: Codec[PaymentHash] = variableSizeBytesLong(varintoverflow, bytes32).as[PaymentHash]
+
+  private val relativeExpiry: Codec[RelativeExpiry] = variableSizeBytesLong(varintoverflow, tu32).as[RelativeExpiry]
+
+  private val cltv: Codec[Cltv] = variableSizeBytesLong(varintoverflow, uint16).as[CltvExpiryDelta].as[Cltv]
+
+  private val fallbackAddress: Codec[FallbackAddress] = variableSizeBytesLong(varintoverflow,
+    ("version" | byte) ::
+      ("address" | variableSizeBytes(uint16, bytes))).as[FallbackAddress]
+
+  private val fallbacks: Codec[Fallbacks] = variableSizeBytesLong(varintoverflow, list(fallbackAddress)).xmap[Seq[FallbackAddress]](_.toSeq, _.toList).as[Fallbacks]
+
+  private val refundSignature: Codec[RefundSignature] = variableSizeBytesLong(varintoverflow, bytes64).as[RefundSignature]
+
+  val invoiceTlvCodec: Codec[TlvStream[InvoiceTlv]] = TlvCodecs.tlvStream[InvoiceTlv](discriminated[InvoiceTlv].by(varint)
+    .typecase(UInt64(3), chain)
+    .typecase(UInt64(4), offerId)
+    .typecase(UInt64(8), amount)
+    .typecase(UInt64(10), description)
+    .typecase(UInt64(12), features)
+    .typecase(UInt64(16), paths)
+    .typecase(UInt64(18), paymentPathsInfo)
+    .typecase(UInt64(19), paymentPathsConstraints)
```

Comment:
Can you add a comment or todo here to mention that we need to work on this at the specification level and it's not final?
We should start with that (we need something), but we'll probably want to re-work that codec once we have agreement on the payment info / constraints mechanism.

Reply:
Done